service/notifier/slack/slack.go

```diff
@@ -0,0 +1,173 @@
+	return Config{
+		// Dependencies.
+		Logger:      nil,
+		SlackClient: nil,
+
+		// Settings.
+		Channel:     "",
+		Emoji:       "",
+		Environment: "",
+		Username:    "",
+	}
+}
+
+// New creates a new configured Slack Notifier.
+func New(config Config) (*Notifier, error) {
+	// Dependencies.
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty")
+	}
+	if config.SlackClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.SlackClient must not be empty")
+	}
+
+	// Settings.
+	if config.Channel == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.Channel must not be empty")
+	}
+	if config.Emoji == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.Emoji must not be empty")
+	}
+	if config.Environment == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.Environment must not be empty")
+	}
+	if config.Username == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.Username must not be empty")
+	}
+
+	// TODO add common health check for this.
+	config.Logger.Log("debug", "checking connection to Slack")
+	_, err := config.SlackClient.AuthTest()
+	if err != nil {
+		return nil, microerror.Maskf(err, "could not authenticate with slack")
+	}
+
+	notifier := &Notifier{
+		// Dependencies.
+		logger:      config.Logger,
+		slackClient: config.SlackClient,
+
+		// Settings.
+		channel:     config.Channel,
+		emoji:       config.Emoji,
+		environment: config.Environment,
+		username:    config.Username,
+	}
+
+	return notifier, nil
+}
+
+// Notifier is an implementation of the Notifier interface,
+// that uses Slack.
+type Notifier struct {
+	// Dependencies.
+	logger      micrologger.Logger
+	slackClient Client
+
+	// Settings.
+	channel     string
+	emoji       string
+	environment string
+	username    string
+}
+
+func (n *Notifier) Failed(project spec.Project, errorMessage string) error {
+	n.logger.Log("debug", "sending failed message to slack")
+	return n.postSlackMessage(project, errorMessage)
+}
+
+func (n *Notifier) Success(project spec.Project) error {
+	n.logger.Log("debug", "sending success message to slack")
+	return n.postSlackMessage(project, "")
+}
+
+// postSlackMessage takes a DeploymentEvent and a possible error message,
+// and posts a helpful message to the configured Slack channel.
+func (n *Notifier) postSlackMessage(project spec.Project, errorMessage string) error {
+	startTime := time.Now()
+	defer updateSlackMetrics(startTime)
+
+	success := false
+	if errorMessage == "" {
+		success = true
+	}
+
+	attachment := slack.Attachment{}
+
+	attachment.Color = dangerColour
+	if success {
+		attachment.Color = goodColour
+	}
+
+	attachment.MarkdownIn = []string{"text"}
+
+	attachment.Title = fmt.Sprintf(titleFormat, project.Name, project.Ref)
+	attachment.Text = fmt.Sprintf(failedMessageFormat, errorMessage)
+	if success {
+		attachment.Text = successMessage
+	}
+	attachment.Footer = n.environment
```

Comment:
The footer misses the deployment event ID. We do not have it yet. We can get this straight in the next step when we add the deployment event ID to the TPR.